telegram/ext/jobqueue.py

```diff
@@ -202,6 +203,114 @@ def run_repeating(self, callback, interval, first=None, context=None, name=None)
+                ``callback.__name__``.
+            day_is_strict (:obj:`bool`, optional): If ``False`` and day > month.days, will pick
+                the last day in the month. Defaults to ``True``.
+
+        Returns:
+            :class:`telegram.ext.Job`: The new ``Job`` instance that has been added to the job
+            queue.
+
+        """
+        if 1 <= day <= 31:
+            next_dt = self._get_next_month_date(day, day_is_strict, when)
+            job = Job(callback, repeat=False, context=context, name=name, job_queue=self,
+                      is_monthly=True, day_is_strict=day_is_strict)
+            self._put(job, time_spec=next_dt)
+            return job
+        else:
+            raise ValueError("The elements of the 'day' argument should be from 1 up to"
+                             " and including 31")
+
+    def _get_next_month_date(self, day, day_is_strict, when):
+        """This method returns the date that the next monthly job should be scheduled.
+
+        Args:
+           day (:obj:`int`): The day of the month the job should run.
+           day_is_strict (:obj:`bool`):
+               Specification as to whether the specified day of job should be strictly
+               respected. If day_is_strict is ``True`` it ignores months whereby the
+               specified date does not exist (e.g February 31st). If it set to ``False``,
+               it returns the last valid date of the month instead. For example,
+               if the user runs a job on the 31st of every month, and sets
+               the day_is_strict variable to ``False``, April, for example,
+               the job would run on April 30th.
+           when (:obj:`datetime.time`): Time of day at which the job should run. If the
+               timezone (``time.tzinfo``) is ``None``, UTC will be assumed.
+
+        """
+        dt = datetime.datetime.now(tz=when.tzinfo or _UTC)
+        dt_time = dt.time().replace(tzinfo=when.tzinfo)
+        if calendar.monthrange(dt.year, dt.month)[1] < day:
+            # if the day does not exist in the current month (e.g Feb 31st)
+            if day_is_strict is False:
+                # set day as last day of month instead
+                next_dt = dt + datetime.timedelta(days=calendar.monthrange(
+                                                  dt.year, dt.month)[1] - dt.day)
+            else:
+                # else set as day in subsequent month. Subsequent month is
+                # guaranteed to have the date, if current month does not have the date.
+                next_dt = dt + datetime.timedelta(days=calendar.monthrange(
+                                                  dt.year, dt.month)[1] - dt.day + day)
+        else:
+            # if the day exists in the current month
+            if dt.day < day:
+                # day is upcoming
+                next_dt = dt + datetime.timedelta(day - dt.day)
+            elif dt.day > day or (dt.day == day and dt_time > when):
```

Comment:
I find this clause is somewhat hard to understand. My suggestion would be:

1. Calculate `next_year` and `next_month` beforehand (with the different cases for December)
2. Use `calendar.monthrange()` to determine `next_month_has_date`
3. Use `next_year` and `next_moth` in the existing `if-else-clauses`. This also removes dubplicate code
4. Add another `else` clause for the case `next_month_has_date is True`

While your code may be a bit more efficient, I think it's beneficial to have it easy to understand in case one has to make changes in the future.

Reply:
sure